### PR TITLE
[move] Remove old core types parser

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

This removes the old type tag parser, and switches us over to always using the new one. This also removes the `from_str` instances for type tag, and struct tag. 

After this there's no parsing logic and code in `move-core-types` 🎉 

## Test plan 

Make sure existing tests pass.

### CODE REVIEW

File: external-crates/move/crates/move-core-types/src/language_storage.rs [hunk 1]
```diff
@@ -133,13 +132,13 @@ impl TypeTag {
...
-        parse_type_tag(s)
-    }
-}
+// impl FromStr for TypeTag {
+//     type Err = anyhow::Error;
+//
+//     fn from_str(s: &str) -> Result<Self, Self::Err> {
+//         parse_type_tag(s)
+//     }
+// }
```
Comment (bmwill) [inline external-crates/move/crates/move-core-types/src/language_storage.rs:141]: So i'm not sure how i feel able this. It seems an easy thing to do to just have the correct parser be in this crate so that we can have a FromStr impl that round-trips with its Display impl so i wonder what your rationale removing these

File: external-crates/move/crates/move-vm-integration-tests/src/tests/loader_tests.rs [hunk 2]
```diff
@@ -62,6 +62,12 @@ struct RelinkingStore {
     type_origin: BTreeMap<(ModuleId, Identifier), ModuleId>,
 }
 
+fn parse_type_tag(s: &str) -> anyhow::Result<TypeTag> {
+    move_command_line_common::types::ParsedType::parse(s)
+        .unwrap()
+        .into_type_tag(&|_| None)
+}
```
Comment (bmwill) [inline external-crates/move/crates/move-vm-integration-tests/src/tests/loader_tests.rs:69]: It just seems a bit clunky to need to define these "parse_type_tag" functions in each crate that wants to do a parse form string -> TypeTag

**Review by dariorussi — COMMENTED** [review]
I guess having the parsing code in `move-command-line-common` feels a bit strange. 
I mean that parsing is pretty core (I think).
Thoughts on moving it elsewhere?
Overall I like having a single parser and I have no problem removing the `from_str` from core code

**Review by amnn — COMMENTED** [review]
Very happy to see the number of parsers for this go down by one -- general +1 to the idea that `move-commandline-common` is a weird place for the canonical parser for these types vs having it in the same crate that defines the type, and having `FromStr` impls (if for no other reason than the very selfish one that this PR is going to create an eye-watering merge conflict with #19732.

**Comment by tzakian** [comment]
Fair points all -- chatted with Ashok for a while this morning and realize I might have been a bit overzealous with the removal of parsing completely from the move core types crate. Will move the source-of-truth parser into there, and will keep `FromStr`. Should have an update for this PR in the next couple of hours.

**Comment by tzakian** [comment]
@amnn @bmwill @dariorussi PR should be updated now so the canonical type tag parser (and module ID/struct tag parsers) are all defined in Move core types.

Did a general cleanup in the process of removing `ParsedType` as that was definitely not going to stick around in move core types :)

File: crates/sui-indexer/README.md [hunk 3]
```diff
@@ -13,7 +13,8 @@ export LDFLAGS="-L/opt/homebrew/opt/postgresql@15/lib"
 export CPPFLAGS="-I/opt/homebrew/opt/postgresql@15/include"
 export PATH="/opt/homebrew/opt/postgresql@15/bin:$PATH"
 ```
-- make sure you have libpq installed: `brew install libpq`, and in your profile, add `export PATH="/opt/homebrew/opt/libpq/bin:$PATH"`. If this doesn't work, try `brew link --force libpq`.
+- make sure you have libpq installed: `brew install libpq`, and in your profile, add `export PATH="/opt/homebrew/opt/libpq/bin:$PATH"`. If this doesn't work, try `brew link --force libpq`. 
+    - Note: Make sure that the `libpq` bin occurs after the `postgres` bin in your `PATH` variable. Otherwise you will get `initdb` errors.
```
Comment (tzakian) [inline crates/sui-indexer/README.md:17]: Added this note as I had issues running tests involving postgres because my `PATH` had the correct elements, but in the wrong order, so it was using the wrong `initdb` in `libpq` as opposed to the installed postgresql@15

File: crates/sui/src/client_ptb/ast.rs [hunk 4]
```diff
@@ -141,6 +140,27 @@ pub enum ParsedPTBCommand {
...
 }
 
+#[derive(Debug, Clone)]
+pub struct ParsedStructType {
+    pub fq_name: ModuleAccess,
+    pub type_args: Vec<ParsedType>,
+}
+
+#[derive(Debug, Clone)]
+pub enum ParsedType {
```
Comment (tzakian) [inline crates/sui/src/client_ptb/ast.rs:150]: The typetag parser no longer does deferred resolution of named addresses. So this needed to be moved into here so that we could support substitution of named addresses based on the current environment when building a PTB.

File: crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs [hunk 5]
```diff
@@ -38,7 +36,7 @@ pub struct ParsedMoveCall {
     pub package: Identifier,
     pub module: Identifier,
     pub function: Identifier,
-    pub type_arguments: Vec<ParsedType>,
+    pub type_arguments: Vec<String>,
```
Comment (tnowacki) [inline crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs:39]: This feels a bit worse to me. Why the change?

File: crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs [hunk 6]
```diff
@@ -284,22 +282,41 @@ where
...
+        let mut start = 0;
+        for (i, c) in contents.char_indices() {
+            match c {
+                '<' => open_pins += 1,
+                '>' => open_pins -= 1,
+                ',' if open_pins == 0 => {
+                    res.push(contents[start..i].to_string());
+                    start = i + 1;
+                }
+                _ => {}
```
Comment (tnowacki) [inline crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs:308]: Not sure what is going on here, but it feels like something we should do with a parser?

File: crates/sui/src/client_ptb/ast.rs (hunk 4 above)
Comment (tnowacki) [reply crates/sui/src/client_ptb/ast.rs:150]: Why not? This was the whole point of having `ParsedType` vs `TypeTag` or any similar form

Comment (tzakian) [reply crates/sui/src/client_ptb/ast.rs:150]: The main reason is a balk at exposing `ParsedType` from `move-core-types`, and if we want to keep `ParsedType` while not having it in `move-core-types` we either need to make that parser the source of truth, and the parser should then be in a different crate than `move-core-types`. However then you can't write `from_str` instances for it and there's a generally fairly strong feeling that the parser for type tags should live in `move-core-types`, which I think is a fair point. 

Generally, simplifying this form so that we resolve while parsing feels simpler to me as we don't need to then build up a bunch of parallel datastructures to `TypeTag` to handle unresolved addresses in types. Additionally, the existing `ParsedType` form was only used for testing code (which can be handled by deferring the parsing of the type until the usage site in the tests), or in the PTB cli case but there we're already parsing into our own form with our own parser anyways so we can capture source info, so there isn't any really advantage to keeping it for that.

File: crates/sui-indexer/README.md (hunk 3 above)
Comment (tzakian) [reply crates/sui-indexer/README.md:17]: I can? But it's so small...

File: crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs (hunk 5 above)
Comment (tzakian) [reply crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs:39]: This is because we are deferring the parsing of the types till the callsite when we have the named addresses in scope. I think I can parse it directly though. Let me take a look!

File: crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs (hunk 6 above)
Comment (tzakian) [reply crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/parser.rs]: All gone -- parsing it in directly now :)

File: crates/sui/src/client_ptb/ast.rs (hunk 4 above)
Comment (tnowacki) [reply crates/sui/src/client_ptb/ast.rs:150]: > The main reason is I'm not a gigantic fan of exposing ParsedType from move-core-types

Why? If you are exposing the parser, whats the harm? 
You can have the `from_str` instances of `TypeTag` operate over an empty mapping of named addresses, it will still then round trip from `Display`

Comment (tzakian) [reply crates/sui/src/client_ptb/ast.rs:150]: I guess to me it's just one more thing that you then need to worry about, and adds unneeded complexity, both to the parser and data structures (now you need to worry about address format, how the different parsed data structures matchup with the type tag, etc), as well as the interface of the crate. 

I would prefer to keep things as tight as possible around type tags and the interface around them, as in the past I think we've offered a bit too much flexibility and different options in this area. Basically, I want to move us into a world where there is only one type of "type" or parsed type as far as move core types is concerned, and that is `TypeTag`. You can go from a string into a `TypeTag`, and you can resolve named addresses if you want while doing that, but that's it, and once you have a `TypeTag` that's it.

Comment (tzakian) [reply crates/sui/src/client_ptb/ast.rs:150]: Might be easier to chat about this synchronously, so let's discuss at the open discussion tomorrow if we have time.

Comment (bmwill) [reply crates/sui/src/client_ptb/ast.rs:150]: Imo named addresses and type tag parsing should likely be handled separately.

**Comment by tzakian** [comment]
Chatted offline and I'm going to split this off into multiple smaller PRs first where we can discuss as they get put up. Closing this.
